pystiche_papers/ulyanov_et_al_2016/_loss.py

```diff
@@ -99,6 +100,22 @@ def content_loss(
     )
 
 
+# Scale gradients in the backward pass
+# https://github.com/jcjohnson/neural-style/issues/450
+# https://github.com/ProGamerGov/neural-style-pt/blob/cbcd023326a3487a2d75270ed1f3b3ddb4b72407/neural_style.py#L404
+class ScaleGradients(torch.autograd.Function):
+    @staticmethod
+    def forward(self: Any, input_tensor: torch.Tensor, weight: float) -> torch.Tensor:  # type: ignore[override]
+        self.weight = weight
+        return input_tensor
+
+    @staticmethod
+    def backward(self: Any, grad_output: torch.Tensor) -> Tuple[torch.Tensor, Any]:  # type: ignore[override]
+        grad_input = grad_output.clone()
+        grad_input = grad_input / (torch.norm(grad_input, keepdim=True) + 1e-8)
+        return grad_input * self.weight, None
```

Comment:
Wouldn't this multiply the weights twice together with L180?

Reply:
They also have a multiplication in the backward pass see [L60](https://github.com/DmitryUlyanov/texture_nets/blob/b2097eccaec699039038970b191780f97c238816/src/texture_loss.lua#L60). However, in the solution of ProGamerGov is a further multiplication of the weight [L414](https://github.com/ProGamerGov/neural-style-pt/blob/cbcd023326a3487a2d75270ed1f3b3ddb4b72407/neural_style.py#L414), but I think that's just a typing there. But I'm just about to read the discussion of it.

Reply:
> They also have a multiplication in the backward pass see [L60](https://github.com/DmitryUlyanov/texture_nets/blob/b2097eccaec699039038970b191780f97c238816/src/texture_loss.lua#L60).

In their implementation the forward and backward pass are independent of each other. That means, the multiplication in the forward pass does not affect the gradient at all. 

In our implementation we only need to define the forward pass and and the backward pass is added automatically. Since we have the multiplication in the forward pass, it is automatically included in the backward pass.

Thus, if you also include it into the scaling, it is performed twice.